### PR TITLE
Adds support for executing locked tests via the API

### PR DESCRIPTION
**Pull request by jathak**
This adds preliminary support for executing locked tests via the API, as requested in #155.

Currently, this works with `DoctestSuite` and the other suites that inherit from it, but not `WwppSuite` or `ConceptSuite` (add support for these should be fairly trivial if necessary, though).

I've manually tested this with one of the 61A projects, but I still need to write some actual tests.

### CODE REVIEW

**Comment by papajohn** [comment]
I think `DoctestSuite` is the only type we need for Data 8. I'd rather have `skip_locked_cases=False` be the default, so that they are evaluated normally. It's often the case in Data 8 (and probably other interactive settings) that unlocking & testing are equivalent, because the test just checks that some name is correctly bound to some value. Hence, we don't plan to much (if any) unlocking in the interactive case.

File: client/sources/common/interpreter.py [hunk 1]
```diff
@@ -251,9 +253,17 @@ def _compare(self, expected, code):
                 actual = (printed + self._output_fn(value)).strip()
             else:
                 actual = printed.strip()
-
+
```
Comment (knrafto) [inline client/sources/common/interpreter.py]: Whitespace?

**Comment by jathak** [comment]
I added a couple of tests. This should now be good to merge (unless anyone sees something I should change).

File: client/sources/common/interpreter.py [hunk 2]
```diff
@@ -253,7 +255,15 @@ def _compare(self, expected, code):
                 actual = printed.strip()
 
         expected = expected.strip()
-        if expected != actual:
+        
+        if not self.skip_locked_cases and expected != actual:
+            actual = hmac.new(self.hash_key.encode('utf-8'),
```
Comment (albert12132) [inline client/sources/common/interpreter.py]: Now that this locking procedure is being used in multiple places, you should move it into its own file into a common directory (probably `client/utils`, since this is used in both `client/sources` and `client/protocols`):

1. Better abstraction: callers of the util won't (and shouldn't) need to know that locking is a hash using `hmac`.
2. Easier to maintain: if we ever change the locking implementation, we would only need to change it in `utils` instead of across several files.

Off the top of my head, I think this locking is being used in LockProtocol, UnlockProtocol, and here. You can also use the new locking util in the tests you wrote.

**Comment by jathak** [comment]
Calls to `hmac` now call a function in `client/utils/locking.py`.

File: demo/ok_test/hw1.py [hunk 3]
```diff
@@ -6,5 +6,5 @@ def square(x):
 
 def double(x):
     """Return x doubled."""
-    return x # Incorrect
+    return x*2 # Incorrect
```
Comment (albert12132) [inline demo/ok_test/hw1.py]: This was left intentionally incorrect.

File: demo/ok_test/tests/q1.py [hunk 4]
```diff
@@ -17,9 +16,10 @@
           'question': 'What is the domain and range of the square function?'
         }
       ],
+      'scored': False,
```
Comment (albert12132) [inline demo/ok_test/tests/q1.py]: It doesn't look like any functional changes were made to this file.

File: demo/ok_test/tests/q2.py [hunk 5]
```diff
@@ -35,6 +35,15 @@
           # explanation: doubling a negative number
           """,
           'hidden': False
+        },
+        {
+          'code': r"""
+          >>> print("hello world") # doctest:+ELLIPSIS
+          [...] world
```
Comment (albert12132) [inline demo/ok_test/tests/q2.py]: Is this related to this PR?

File: client/protocols/unlock.py [hunk 6]
```diff
@@ -173,8 +172,7 @@ def interact(self, unique_id, case_id, question_prompt, answer, choices=None, ra
     ###################
 
     def _verify(self, guess, locked):
-        return hmac.new(self.hash_key.encode('utf-8'),
-                        guess.encode('utf-8')).hexdigest() == locked
+        return locking.lock(self.hash_key, guess).hexdigest() == locked
```
Comment (albert12132) [inline client/protocols/unlock.py]: The `.hexdigest()` here isn't needed, since `locking.lock` already does it.

File: client/protocols/lock.py [hunk 7]
```diff
@@ -1,6 +1,5 @@
 from client.protocols.common import models
-from client.utils import format
-import hmac
+from client.utils import format, locking
```
Comment (albert12132) [inline client/protocols/lock.py]: Please import each module on its own line.

**Comment by jathak** [comment]
Whoops! I didn't mean to commit those changes to the demo files. I've now reverted those files and made the other changes you commented about.

**Comment by albert12132** [comment]
LGTM. Feel free to merge when you're ready.
